File: swecli/ui_textual/autocomplete_internal/gitignore.py
"""GitIgnore parser utility for autocomplete filtering."""

import os
from pathlib import Path
from typing import Optional

import pathspec
from pathspec import PathSpec


class GitIgnoreParser:
    """Parser for .gitignore files that supports nested gitignore files.

    This class loads and parses .gitignore files from a root directory
    and all its subdirectories, providing methods to check if paths
    should be ignored based on gitignore patterns.
    """

    # Always ignore these directories regardless of .gitignore
    # These are obviously generated/system dirs that are NEVER useful to autocomplete
    ALWAYS_IGNORE_DIRS = {
        # Version Control
        ".git", ".hg", ".svn", ".bzr", "_darcs", ".fossil",

        # OS Generated
        ".DS_Store", ".Spotlight-V100", ".Trashes",
        "Thumbs.db", "desktop.ini", "$RECYCLE.BIN",

        # Python caches
        "__pycache__", ".pytest_cache", ".mypy_cache", ".pytype", ".pyre",
        ".hypothesis", ".tox", ".nox", "cython_debug", ".eggs",

        # Node/JS caches
        "node_modules", ".npm", ".yarn", ".pnpm-store",
        ".next", ".nuxt", ".output", ".svelte-kit", ".angular",
        ".parcel-cache", ".turbo",

        # IDE/Editor
        ".idea", ".vscode", ".vs", ".settings",

        # Java/Kotlin
        ".gradle",

        # Elixir
        "_build", "deps", ".elixir_ls",

        # iOS
        "Pods", "DerivedData", "xcuserdata",

        # Ruby
        ".bundle",

        # Virtual Environments
        ".venv", "venv",

        # Misc caches
        ".cache", ".sass-cache", ".eslintcache", ".stylelintcache",
        ".tmp", ".temp", "tmp", "temp",
    }

    def __init__(self, root_dir: Path):
        """Initialize the GitIgnore parser.

        Args:
            root_dir: Root directory to start parsing from
        """
        self.root_dir = root_dir.resolve()
        # List of (directory, PathSpec) tuples - directory is where the .gitignore lives
        self._specs: list[tuple[Path, PathSpec]] = []
        self._load_gitignore_files()

    def _load_gitignore_files(self) -> None:
        """Find and parse all .gitignore files in the directory tree."""
        # First, load root .gitignore if it exists
        root_gitignore = self.root_dir / ".gitignore"
        if root_gitignore.exists():
            spec = self._parse_gitignore(root_gitignore)
            if spec:
                self._specs.append((self.root_dir, spec))

        # Walk directory tree to find nested .gitignore files
        # We need to be careful not to descend into always-ignored directories
        try:
            for root, dirs, files in os.walk(self.root_dir):
                # Skip always-ignored directories
                dirs[:] = [d for d in dirs if d not in self.ALWAYS_IGNORE_DIRS]

                root_path = Path(root)

                # Skip the root directory since we already processed it
                if root_path == self.root_dir:
                    continue

                # Check for .gitignore in this directory
                gitignore_path = root_path / ".gitignore"
                if gitignore_path.exists():
                    spec = self._parse_gitignore(gitignore_path)
                    if spec:
                        self._specs.append((root_path, spec))
        except (PermissionError, OSError):
            pass

    def _parse_gitignore(self, gitignore_path: Path) -> Optional[PathSpec]:
        """Parse a single .gitignore file.

        Args:
            gitignore_path: Path to the .gitignore file

        Returns:
            PathSpec object or None if parsing fails
        """
        try:
            with open(gitignore_path, "r", encoding="utf-8", errors="ignore") as f:
                lines = f.readlines()

            # Filter out comments and empty lines, normalize patterns
            patterns = []
            for line in lines:
                line = line.strip()
                # Skip empty lines and comments
                if not line or line.startswith("#"):
                    continue
                patterns.append(line)

            if patterns:
                return pathspec.PathSpec.from_lines(
                    pathspec.patterns.GitWildMatchPattern, patterns
                )
        except (IOError, OSError):
            pass

        return None

    def is_ignored(self, path: Path) -> bool:
        """Check if a path should be ignored based on gitignore patterns.

        Args:
            path: Absolute or relative path to check

        Returns:
            True if the path matches any gitignore pattern
        """
        # Ensure we have an absolute path
        if not path.is_absolute():
            path = self.root_dir / path

        path = path.resolve()

        # Check if path is under root
        try:
            rel_path = path.relative_to(self.root_dir)
        except ValueError:
            return False

        # Check against always-ignored directories
        for part in rel_path.parts:
            if part in self.ALWAYS_IGNORE_DIRS:
                return True

        # Check against each applicable gitignore spec
        for spec_dir, spec in self._specs:
            # Only apply specs that are in the path's ancestor chain
            try:
                # Check if path is under spec_dir
                path.relative_to(spec_dir)
            except ValueError:
                # Path is not under this spec's directory
                continue

            # Get path relative to spec's directory for matching
            try:
                match_path = path.relative_to(spec_dir)
            except ValueError:
                continue

            # Normalize path for matching
            match_str = str(match_path).replace(os.sep, "/")

            # Add trailing slash for directories
            if path.is_dir() and not match_str.endswith("/"):
                match_str += "/"

            if spec.match_file(match_str):
                return True

        return False

    def should_skip_dir(self, dir_path: Path) -> bool:
        """Check if a directory should be skipped during traversal.

        This is a convenience method for use with os.walk() that checks
        if a directory should be pruned from the traversal.

        Args:
            dir_path: Path to the directory

        Returns:
            True if the directory should be skipped
        """
        dir_name = dir_path.name

        # Always skip these directories
        if dir_name in self.ALWAYS_IGNORE_DIRS:
            return True

        return self.is_ignored(dir_path)

    def filter_paths(self, paths: list[Path]) -> list[Path]:
        """Filter a list of paths, removing ignored ones.

        Args:
            paths: List of paths to filter

        Returns:
            List of paths that are not ignored
        """
        return [p for p in paths if not self.is_ignored(p)]

File: crates/opendev-runtime/src/gitignore.rs
//! GitIgnore parser — filter files based on .gitignore patterns.
//!
//! Parses `.gitignore` files from a root directory and all subdirectories,
//! supporting nested gitignore overrides. Always ignores common directories
//! (`.git`, `node_modules`, `__pycache__`, etc.) regardless of patterns.

use std::path::{Path, PathBuf};

use tracing::debug;

/// Directories always ignored regardless of `.gitignore` contents.
pub const ALWAYS_IGNORE_DIRS: &[&str] = &[
    // Version control
    ".git",
    ".hg",
    ".svn",
    ".bzr",
    "_darcs",
    ".fossil",
    // OS generated
    ".DS_Store",
    ".Spotlight-V100",
    ".Trashes",
    "Thumbs.db",
    "desktop.ini",
    "$RECYCLE.BIN",
    // Python caches
    "__pycache__",
    ".pytest_cache",
    ".mypy_cache",
    ".pytype",
    ".pyre",
    ".hypothesis",
    ".tox",
    ".nox",
    "cython_debug",
    ".eggs",
    // Node/JS caches
    "node_modules",
    ".npm",
    ".yarn",
    ".pnpm-store",
    ".next",
    ".nuxt",
    ".output",
    ".svelte-kit",
    ".angular",
    ".parcel-cache",
    ".turbo",
    // IDE/Editor
    ".idea",
    ".vscode",
    ".vs",
    ".settings",
    // Java/Kotlin
    ".gradle",
    // Elixir
    "_build",
    ".elixir_ls",
    // iOS
    "Pods",
    "DerivedData",
    "xcuserdata",
    // Ruby
    ".bundle",
    // Virtual environments
    ".venv",
    "venv",
    // Misc caches
    ".cache",
    ".sass-cache",
    ".eslintcache",
    ".stylelintcache",
    ".tmp",
    ".temp",
    "tmp",
    "temp",
    // Rust
    "target",
];

/// A parsed `.gitignore` pattern.
#[derive(Debug, Clone)]
struct GitIgnorePattern {
    /// The raw pattern string.
    pattern: String,
    /// Whether this is a negation pattern (starts with `!`).
    negated: bool,
    /// Whether this only matches directories (ends with `/`).
    dir_only: bool,
}

/// A `.gitignore` spec loaded from a specific directory.
#[derive(Debug, Clone)]
struct GitIgnoreSpec {
    /// Directory where this `.gitignore` was found.
    base_dir: PathBuf,
    /// Parsed patterns from the file.
    patterns: Vec<GitIgnorePattern>,
}

/// GitIgnore parser that supports nested `.gitignore` files.
pub struct GitIgnoreParser {
    root_dir: PathBuf,
    specs: Vec<GitIgnoreSpec>,
}

impl GitIgnoreParser {
    /// Create a new parser rooted at the given directory.
    pub fn new(root_dir: &Path) -> Self {
        let root_dir = root_dir
            .canonicalize()
            .unwrap_or_else(|_| root_dir.to_path_buf());
        let mut parser = Self {
            root_dir,
            specs: Vec::new(),
        };
        parser.load_gitignore_files();
        parser
    }

    /// Check whether a path should be ignored.
    pub fn is_ignored(&self, path: &Path) -> bool {
        let abs_path = if path.is_absolute() {
            path.to_path_buf()
        } else {
            self.root_dir.join(path)
        };

        let rel = match abs_path.strip_prefix(&self.root_dir) {
            Ok(r) => r,
            Err(_) => return false,
        };

        // Check always-ignored directories
        for component in rel.components() {
            let s = component.as_os_str().to_string_lossy();
            if ALWAYS_IGNORE_DIRS.contains(&s.as_ref()) {
                return true;
            }
        }

        // Check gitignore patterns
        let mut ignored = false;
        for spec in &self.specs {
            // Only apply spec if path is under spec's base dir
            let spec_rel = match abs_path.strip_prefix(&spec.base_dir) {
                Ok(r) => r,
                Err(_) => continue,
            };

            let match_str = spec_rel.to_string_lossy().replace('\\', "/");
            let is_dir = abs_path.is_dir();

            for pat in &spec.patterns {
                if pat.dir_only && !is_dir {
                    continue;
                }

                if matches_pattern(&pat.pattern, &match_str) {
                    ignored = !pat.negated;
                }
            }
        }

        ignored
    }

    /// Check if a directory name is in the always-ignore list.
    pub fn is_always_ignored(name: &str) -> bool {
        ALWAYS_IGNORE_DIRS.contains(&name)
    }

    fn load_gitignore_files(&mut self) {
        // Load root .gitignore
        let root_gitignore = self.root_dir.join(".gitignore");
        if root_gitignore.exists()
            && let Some(spec) = self.parse_gitignore(&root_gitignore, &self.root_dir.clone())
        {
            self.specs.push(spec);
        }

        // Walk subdirectories
        self.walk_for_gitignores(&self.root_dir.clone());
    }

    fn walk_for_gitignores(&mut self, dir: &Path) {
        let entries = match std::fs::read_dir(dir) {
            Ok(e) => e,
            Err(_) => return,
        };

        for entry in entries.flatten() {
            let path = entry.path();
            if !path.is_dir() {
                continue;
            }

            let name = path
                .file_name()
                .map(|n| n.to_string_lossy().to_string())
                .unwrap_or_default();

            if ALWAYS_IGNORE_DIRS.contains(&name.as_str()) {
                continue;
            }

            let gitignore = path.join(".gitignore");
            if gitignore.exists()
                && let Some(spec) = self.parse_gitignore(&gitignore, &path)
            {
                self.specs.push(spec);
            }

            self.walk_for_gitignores(&path);
        }
    }

    fn parse_gitignore(&self, gitignore_path: &Path, base_dir: &Path) -> Option<GitIgnoreSpec> {
        let content = std::fs::read_to_string(gitignore_path).ok()?;
        let mut patterns = Vec::new();

        for line in content.lines() {
            let trimmed = line.trim();
            if trimmed.is_empty() || trimmed.starts_with('#') {
                continue;
            }

            let (pattern, negated) = if let Some(rest) = trimmed.strip_prefix('!') {
                (rest.to_string(), true)
            } else {
                (trimmed.to_string(), false)
            };

            let dir_only = pattern.ends_with('/');
            let pattern = if dir_only {
                pattern.trim_end_matches('/').to_string()
            } else {
                pattern
            };

            patterns.push(GitIgnorePattern {
                pattern,
                negated,
                dir_only,
            });
        }

        if patterns.is_empty() {
            debug!("No patterns in {}", gitignore_path.display());
            return None;
        }

        Some(GitIgnoreSpec {
            base_dir: base_dir.to_path_buf(),
            patterns,
        })
    }
}

/// Simple glob pattern matching (supports `*`, `**`, `?`).
fn matches_pattern(pattern: &str, path: &str) -> bool {
    // Handle patterns starting with `/` (root-relative)
    let pattern = pattern.strip_prefix('/').unwrap_or(pattern);

    if pattern.contains("**") {
        // ** matches any number of directories
        let parts: Vec<&str> = pattern.split("**").collect();
        if parts.len() == 2 {
            let prefix = parts[0].trim_end_matches('/');
            let suffix = parts[1].trim_start_matches('/');
            if prefix.is_empty() && suffix.is_empty() {
                return true;
            }
            if prefix.is_empty() {
                return path.ends_with(suffix) || simple_match(suffix, path);
            }
            if suffix.is_empty() {
                return path.starts_with(prefix) || simple_match(prefix, path);
            }
            // Check if path starts with prefix and ends with suffix
            return path.contains(prefix) && path.contains(suffix);
        }
    }

    // If pattern has no slash, it matches any file with that name
    if !pattern.contains('/') {
        // Match against the last component
        let file_name = path.rsplit('/').next().unwrap_or(path);
        return simple_match(pattern, file_name) || simple_match(pattern, path);
    }

    simple_match(pattern, path)
}

/// Simple wildcard matching (`*` matches anything except `/`, `?` matches one char).
fn simple_match(pattern: &str, text: &str) -> bool {
    let p: Vec<char> = pattern.chars().collect();
    let t: Vec<char> = text.chars().collect();
    simple_match_impl(&p, &t)
}

fn simple_match_impl(pattern: &[char], text: &[char]) -> bool {
    if pattern.is_empty() {
        return text.is_empty();
    }
    if pattern[0] == '*' {
        // Skip consecutive stars
        let mut i = 0;
        while i < pattern.len() && pattern[i] == '*' {
            i += 1;
        }
        if i >= pattern.len() {
            return true;
        }
        for j in 0..=text.len() {
            if simple_match_impl(&pattern[i..], &text[j..]) {
                return true;
            }
        }
        return false;
    }
    if text.is_empty() {
        return false;
    }
    if pattern[0] == '?' || pattern[0] == text[0] {
        return simple_match_impl(&pattern[1..], &text[1..]);
    }
    false
}

impl std::fmt::Debug for GitIgnoreParser {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("GitIgnoreParser")
            .field("root_dir", &self.root_dir)
            .field("specs_count", &self.specs.len())
            .finish()
    }
}

#[cfg(test)]
#[path = "gitignore_tests.rs"]
mod tests;
